server/config/config.go

```diff
@@ -474,19 +546,26 @@ func parseConfig(fileBytes []byte) (*generalConfig, error) {
 		log.Warningf("Unknown fields in config: %s", err)
 	}
 
-	if err := yaml.Unmarshal([]byte(expandedFileBytes), &sharedGeneralConfig); err != nil {
+	gc := &generalConfig{}
+	if err := yaml.Unmarshal([]byte(expandedFileBytes), gc); err != nil {
 		return nil, fmt.Errorf("Error parsing config file: %s", err)
 	}
+	// The shared config caches the last config parsed from a file so that a call to
+	// `readConfig` with an empty `configFile` can return the cached config.
+	sharedGeneralConfig = gc
 
-	// Re-parse flags so that they override the YAML config values.
-	flag.Parse()
+	generalConfigMap := make(map[interface{}]interface{})
+	if err := yaml.Unmarshal([]byte(expandedFileBytes), &generalConfigMap); err != nil {
+		return nil, fmt.Errorf("Error parsing config file: %s", err)
+	}
+	populateYamlPresenceMap(generalConfigMap, presenceMap)
 
-	return &sharedGeneralConfig, nil
+	return gc, nil
 }
 
-func readConfig(fullConfigPath string) (*generalConfig, error) {
+func readConfig(fullConfigPath string, presenceMap map[string]struct{}) (*generalConfig, error) {
```

Comment:
Is it necessary to pass the presence map as an out parameter? Could you instantiate the empty map inside the function and return it instead?